File: prefix_sum/shortest_subarray_with_sum_at_least_k/solution2.go
func shortestSubarray(nums []int, k int) int {
    prefixSums, sum := []int{0}, 0 // sumof[i...j] = prefixSums[j+1]-prefixSum[i]
	for i := 0; i < len(nums); i++ {
		sum += nums[i]
		prefixSums = append(prefixSums, sum)
	}

	monoQ, ret := make([]int, 0), math.MaxInt
	for j, prefixSum := range prefixSums {
		for ; len(monoQ) > 0 && prefixSum-prefixSums[monoQ[0]] >= k; ret, monoQ = min(ret, j-monoQ[0]), monoQ[1:] {
		}

		for ; len(monoQ) > 0 && prefixSum <= prefixSums[monoQ[len(monoQ)-1]]; monoQ = monoQ[:len(monoQ)-1] {
		}
		monoQ = append(monoQ, j)
	}
	if ret == math.MaxInt {
		ret = -1
	}

	return ret
}

/*
	执行用时分布：14ms，击败 35.44%
	消耗内存分布：10.07MB，击败 63.29%
*/

File: binary_search/find_minimum_in_rotated_sorted_array/solution2.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return 10000

        m = (len(nums)-1)//2
        if nums[0] <= nums[m]:
            return min(nums[0], self.findMin(nums[m+1:]))
        else:
            return min(self.findMin(nums[:m+1]), nums[m+1])    

# 执行用时分布：0ms 击败 100.00%
# 消耗内存分布：19.28MB，击败 27.06%
    